src/main/java/picard/illumina/CollectIlluminaBasecallingMetrics.java

```diff
@@ -178,38 +195,94 @@ protected int doWork() {
+
+            //CBCLs
+            List<File> cbcls = new ArrayList<>();
+            Arrays.asList(cycleDirs)
+                    .forEach(cycleDir -> cbcls.addAll(
+                            Arrays.asList(IOUtil.getFilesMatchingRegexp(
+                                    cycleDir, "^" + IlluminaFileUtil.longLaneStr(LANE) + "_(\\d{1,5}).cbcl$"))));
+
+            if (cbcls.size() == 0) {
+                throw new PicardException("No CBCL files found.");
+            }
+
+            IOUtil.assertFilesAreReadable(cbcls);
+
+            //locs
+            List<AbstractIlluminaPositionFileReader.PositionInfo> locs = new ArrayList<>();
+            File locsFile = new File(BASECALLS_DIR.getParentFile(), "s.locs");
+            try (LocsFileReader locsFileReader = new LocsFileReader(locsFile)) {
+                while (locsFileReader.hasNext()) {
+                    locs.add(locsFileReader.next());
+                }
+            }
+            IOUtil.assertFileIsReadable(locsFile);
+
+            //filter
+            Pattern laneTileRegex = Pattern.compile(ParameterizedFileUtil.escapePeriods(
+                    ParameterizedFileUtil.makeLaneTileRegex(".filter", LANE)));
+            File[] filterFiles = getTiledFiles(laneDir, laneTileRegex);
+
+            IOUtil.assertFilesAreReadable(Arrays.asList(filterFiles));
+            Pattern barcodeRegex = Pattern.compile(ParameterizedFileUtil.escapePeriods(
+                    ParameterizedFileUtil.makeLaneTileRegex("_barcode.txt", LANE)));
+            Map<Integer, File> barcodesFiles = new HashMap<>();
+            for (File barcodeFile : getTiledFiles(laneDir, barcodeRegex)) {
+                Matcher tileMatcher = barcodeRegex.matcher(barcodeFile.getName());
+                if (tileMatcher.matches()) {
+                    barcodesFiles.put(Integer.valueOf(tileMatcher.group(1)), barcodeFile);
+                }
+            }
 
-        while (provider.hasNext()) {
-            final ClusterData cluster = provider.next();
-            addCluster(cluster);
+            factory.getAvailableTiles().forEach(tile -> {
+                File barcodeFile = barcodesFiles.get(tile);
+                BaseIlluminaDataProvider provider = factory.makeDataProvider(cbcls, locs, filterFiles, tile, barcodeFile);
+                while (provider.hasNext()) {
+                    addCluster(provider.next());
+                }
+            });
+        } else {
+            BaseIlluminaDataProvider provider = factory.makeDataProvider();
+
+            while (provider.hasNext()) {
+                final ClusterData cluster = provider.next();
+                addCluster(cluster);
+            }
         }
 
+
         onComplete();
         return 0;
     }
 
     /***
      * Process new cluster of Illumina data - increment a running counter of data
      */
-    private void addCluster(final ClusterData cluster) {
+    private synchronized void addCluster(final ClusterData cluster) {
```

Comment:
Actually not needed.. I was testing with a parellell stream but that proved unfruitful